torch/csrc/Module.cpp

```diff
@@ -1095,6 +1096,224 @@ PyObject* THPModule_getDefaultDevice(PyObject* _unused, PyObject* arg) {
   END_HANDLE_TH_ERRORS
 }
 
+PyObject* THPModule_getAccelerator(PyObject* _unused, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  const auto device_type = at::getAccelerator(false);
+  if (device_type.has_value()) {
+    return THPUtils_packString(c10::DeviceTypeName(
+        device_type.value(),
+        /*lower_case=*/true));
+  } else {
+    Py_RETURN_NONE;
+  }
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* THPModule_hasAccelerator(PyObject* _unused, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  const auto device_type = at::getAccelerator(false);
+  if (!device_type.has_value()) {
+    Py_RETURN_FALSE;
+  }
+  torch::utils::maybe_initialize_device(device_type.value());
+  c10::impl::VirtualGuardImpl impl(device_type.value());
+  if (impl.deviceCount() > 0) {
+    Py_RETURN_TRUE;
+  }
+  Py_RETURN_FALSE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* THPModule_deviceCount(PyObject* _unused, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  const auto device_type = at::getAccelerator(false);
+  if (!device_type.has_value()) {
+    return THPUtils_packUInt64(0);
+  }
+  torch::utils::maybe_initialize_device(device_type.value());
+  c10::impl::VirtualGuardImpl impl(device_type.value());
+  return THPUtils_packDeviceIndex(impl.deviceCount());
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* THPModule_setDevice(
+    PyObject* _unused,
+    PyObject* args,
+    PyObject* kwargs) {
+  HANDLE_TH_ERRORS
+  static torch::PythonArgParser parser({
+      "set_device(Device device)",
+  });
+  torch::ParsedArgs<1> parsed_args{};
+  auto r = parser.parse(args, kwargs, parsed_args);
+  auto device = r.device(0);
+  c10::DeviceType device_type = device.type();
+  TORCH_CHECK(
+      at::isAccelerator(device_type),
+      c10::DeviceTypeName(device_type),
+      " device type is not an accelerator");
+  TORCH_CHECK(
+      device.has_index(),
+      "Expected a torch.device with a specified index or an integer.");
+  // If device index is negative, no-op
+  if (device.index() < 0) {
+    Py_RETURN_NONE;
+  }
+  torch::utils::maybe_initialize_device(device_type);
+  c10::impl::VirtualGuardImpl impl(device_type);
+  impl.setDevice(device);
+  Py_RETURN_NONE;
+  END_HANDLE_TH_ERRORS
+}
+
+PyObject* THPModule_getDevice(PyObject* _unused, PyObject* arg) {
+  HANDLE_TH_ERRORS
+  const auto device_type = at::getAccelerator(true).value();
```

Comment:
Is it expected that the `getDevice` throws an exception because of `at::getAccelerator(true)`?

Reply:
I think yes. We should expect the exception when the user calls an accelerator runtime API on the machine accelerator is not available.